src/store/helpers.ts

```diff
@@ -18,3 +23,50 @@
 export function createState<T>(state: T, partial: Partial<T>): T {
   return Object.assign({}, state, partial);
 }
+
+/**
+ * Returns a devfile from the FactoryResolver object.
+ * @param data a FactoryResolver object.
+ */
+export function getDevfile(data: FactoryResolver): api.che.workspace.devfile.Devfile {
+  let devfile = data.devfile;
+
+  if (isDevfileV2(devfile)) {
+    // temporary solution for fix che-server serialisation bug with empty volume
+    const components = devfile.components.map(component => {
+      if (Object.keys(component).length === 1 && component.name) {
+        component.volume = {};
+      }
+      return component;
+    }) || [];
+    devfile = Object.assign(devfile, { components });
+    // add a default project
+    const projects: ProjectV2Source[] = [];
+    const scmInfo = data['scm_info'];
+    if (scmInfo) {
+      const origin = scmInfo['clone_url'];
+      const name = getProjectName(origin);
+      const revision = scmInfo.branch;
+      const project: ProjectV2Source = { name, git: { remotes: { origin } } };
+      if (revision) {
+        project.git.checkoutFrom = { revision };
+      }
+      projects.push(project);
```

Comment:
well, Do I miss check or no if statement exists that we provision project only in case none is specified in namespace?

Reply:
Yes, if statement exist it will be overridden: 
```devfile = Object.assign({ projects }, devfile);```

Reply:
And I have improved it
```
...
if (!devfile.projects?.length && scmInfo) {
...
```